dns_server/src/server/mod.rs

```
@@ -0,0 +1,286 @@
+            RrKey::new(ns_rec.name().clone().into(), ns_rec.record_type()),
+            Arc::new(ns_rec),
+        );
+        new_records.insert(
+            RrKey::new(ipv4_rec.name().clone().into(), ipv4_rec.record_type()),
+            Arc::new(ipv4_rec),
+        );
+        new_records.insert(
+            RrKey::new(ipv6_rec.name().clone().into(), ipv6_rec.record_type()),
+            Arc::new(ipv6_rec),
+        );
+        *self.inner.records_mut().await = new_records;
+    }
+}
+
+#[async_trait::async_trait]
+impl Authority for AuthorityImpl {
+    type Lookup = AuthLookup;
+
+    fn zone_type(&self) -> ZoneType {
+        self.inner.zone_type()
+    }
+
+    fn is_axfr_allowed(&self) -> bool {
+        self.inner.is_axfr_allowed()
+    }
+
+    async fn update(&self, update: &MessageRequest) -> UpdateResult<bool> {
+        self.inner.update(update).await
+    }
+
+    fn origin(&self) -> &LowerName {
+        self.inner.origin()
+    }
+
+    async fn lookup(
+        &self,
+        name: &LowerName,
+        query_type: RecordType,
+        lookup_options: LookupOptions,
+    ) -> Result<Self::Lookup, LookupError> {
+        self.refresh().await;
+        self.inner.lookup(name, query_type, lookup_options).await
+    }
+
+    async fn search(
+        &self,
+        request_info: RequestInfo<'_>,
+        lookup_options: LookupOptions,
+    ) -> Result<Self::Lookup, LookupError> {
+        self.refresh().await;
+        self.inner.search(request_info, lookup_options).await
+    }
+
+    async fn get_nsec_records(
+        &self,
+        name: &LowerName,
+        lookup_options: LookupOptions,
+    ) -> Result<Self::Lookup, LookupError> {
+        self.inner.get_nsec_records(name, lookup_options).await
+    }
+}
+
+fn handle_command(auth: &AuthorityImpl, command: ServerCommands) {
+    match command {
+        ServerCommands::AddAddress(IpAddr::V4(ip)) => {
+            auth.ip4.lock().expect("mutex must be valid").push(ip);
```

These expects are too aggressive... deeming the whole program dead just because a wrong format was submitted. Perhaps we should just return an error with one shot.